### ISSUE TITLE
[BUG] maskers transform() silently truncates cleaned signal to the source image's integer dtype

### ISSUE THREAD

**Issue by Remi-Gau** [nilearn/nilearn#6525]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Operating system

- [x] Linux

### Operating system version

Linux Ubuntu (kernel 7.0.0-29-generic)

### Python version

- [x] 3.10

### nilearn version

Dev version (main / 0.14.1.dev, also reproduces on `main` as of commit `7c73c85d9a49c0e60917099e8ace1f41c2dee9c7` and later). Does **not** reproduce on 0.13.1.

### Expected behavior

When `NiftiMasker(..., standardize="zscore_sample")` (or `"psc"`, or `standardize=True`) is used to `transform()` an image whose header dtype is an integer type (e.g. `int16`, common for raw/unprocessed BOLD data), the returned 2D signal array should keep the floating-point precision produced by standardization.

### Current behavior & error messages

`NiftiMasker.transform()` silently casts its output back to the **input image's own header dtype** whenever `NiftiMasker(dtype=...)` is left at its default (`None`). For an `int16` source image, this truncates the continuous z-scored/PSC signal down to a handful of integer levels (e.g. `{-2, -1, 0, 1, 2}`), destroying almost all of the signal's precision. No warning is raised.

This was introduced in #5511 ("[FIX] ensure image estimator can fit and preserve several dtype"), in `nilearn/maskers/nifti_masker.py::NiftiMasker.transform_single_imgs`:

```python
target_dtype = get_target_dtype(img_data_dtype(imgs), self.dtype)
if target_dtype is None:
    target_dtype = img_data_dtype(imgs)

return data.astype(target_dtype)
```

`get_target_dtype` returns `None` specifically to mean "no conversion needed" (per its own docstring: *"If None, no conversion is required"*), but the `if target_dtype is None:` branch overrides that `None` with the original image's raw dtype and casts unconditionally. This is presumably meant to mirror the `dtype`-preservation logic added for `inverse_transform` (`BaseMasker._set_inverse_transform_output_dtype`) elsewhere in the same PR, but applying it to the forward `transform()` path is incorrect: `transform()` output for a masker with `standardize` enabled is fundamentally a floating point quantity and should not be coerced back into the source image's discrete dtype.

I found this while comparing the output of `examples/02_decoding/plot_haxby_full_analysis.py` between the 0.13.1 release and current `main`: decoding AUC scores dropped noticeably across the board (e.g. mask_vt/chair: 0.93 → 0.78) purely because of this silent precision loss feeding into the SVM classifier, not because of any change to the example or the classifier itself. I bisected the regression down to commit `7c73c85d9a49c0e60917099e8ace1f41c2dee9c7` (#5511) using `git bisect` with a fixed dependency environment, and confirmed dependency-version differences and the unrelated `svc_l1`/`LinearSVC(dual=...)` change in the same time range are not responsible.

### Steps and code to reproduce bug

```python
import numpy as np
import nibabel as nib

from nilearn.maskers import NiftiMasker

rng = np.random.default_rng(0)
data = (rng.standard_normal((10, 10, 10, 20)) * 1000).astype(np.int16)
img = nib.Nifti1Image(data, affine=np.eye(4))

mask_data = np.zeros((10, 10, 10), dtype=np.int8)
mask_data[3:7, 3:7, 3:7] = 1
mask_img = nib.Nifti1Image(mask_data, affine=np.eye(4))

masker = NiftiMasker(mask_img=mask_img, standardize="zscore_sample")
X = masker.fit_transform(img)

print("dtype:", X.dtype)  # int16 -- should be float
print("n unique values in column 0:", len(np.unique(X[:, 0])))  # only 2-5 -- should be close to 20 (n_samples)
print(X[:, 0])
```

Output on current `main`:

```
dtype: int16
n unique values in column 0: 3
[ 0  0  0  1 -1  0 -1  0  0  1 -1 -1  0  0  1  0  0  0  1 -1]
```

Expected (and what 0.13.1 produces): `dtype: float32` (or `float64`) with ~20 distinct z-scored values.

**Comment by Remi-Gau** [comment]
actually the bug could appear for anything that can modify the data at transform time (detrend, smoothing, confoun removal, filtering...)

**Comment by Remi-Gau** [comment]
Maybe I am paranoid but this bug seems like it may justify to yank 0.14.0 from pypi.

All it takes is input data stored as int and "some" cleaning done on the data.

**Comment by bthirion** [comment]
You mean, release a 0.14.1 ?

**Comment by emdupre** [comment]
> Maybe I am paranoid but this bug seems like it may justify to yank 0.14.0 from pypi.

I would lightly support this ; I agree we would need to cut 0.14.1 to move forward

**Comment by Remi-Gau** [comment]
> You mean, release a 0.14.1 ?

- yeah realease 0.14.1 VERY soon once #6526 is merged and then remove 0.14.0 from pypi
 
@nilearn/core
